### ISSUE TITLE
Cannot js/require ffi module

### ISSUE THREAD

**Issue by kanaka** [anmonteiro/lumo#67]
Lumo example of the problem:
```clojure
cljs.user=> (js/require "ffi")
TypeError: Cannot assign to read only property 'name' of object '[object Object]'
```

This is with ffi-2.2.0 installed in `node_modules/`. Using node (7.3.0) in the same directory is able to import the module.

I thought this might be a problem with compiled node modules but it works correctly with opencv (which is a fairly large compiled module).

Instrumenting the modules, I was able to track this down to the following line in the `ref` module:

```js
exports.refType = function refType (type) {
...
  if (_type.name) {
    rtn.name = _type.name + '*'
  }
...
}
```

The error occurs when the type passed to refType is [StructType]. With pure node, setting the `name` attribute of this type works fine, but inside lumo, it appears the `name` field of [StructType] is marked read-only for some reason.

**Comment by kanaka** [comment]
Strangely, it appears that the name field in marked read-only in both cases but for some reason in the node case it allows it to be modified. Perhaps a node 7.3 vs node 7.4 difference (with the latter now enforcing)?

**Comment by kanaka** [comment]
I tested with node 7.4.0 and it imports the same ffi module just fine so it does appear to be something unique to the lumo build/packaging.

**Comment by anmonteiro** [comment]
Thanks for reporting this issue, and sorry for my late reply.

This may be something related to the Lumo packaging, because of the V8 startup snapshot that we generate.

I may have had this problem in Node itself, and could only work around it by monkey-patching a part of Node during the snapshot generation process, but I'm not sure if it's even the same problem.

Another thing to check (which I won't really have time to in the next few weeks) is if simply removing some of [these build options](https://github.com/anmonteiro/lumo/blob/master/scripts/package.js#L49-L53) would solve the issue.

**Comment by kanaka** [comment]
Okay, I'm going to try building a non-nexe node 7.4.0 build with those options and see if it has the same problem to try and bifurcate where the problem lies.

**Comment by kanaka** [comment]
I built a generic 7.4.0 node from source using those four arguments and it is able to load the ffi module. Next step, use nexe to build a test the following JS program:

```
console.log("Starting load of module " + process.argv[2])
mod = require(process.argv[2])
console.log(mod)
```

That should tell us if it's something nexe is doing or if it's something the JS/ClojureScript code is doing.

**Comment by kanaka** [comment]
Okay, so using basically the same package.js script and a mostly empty snapshot file, I was able to build a simple executable that allows me to pass a module on the command line to load and then prints out the contents. This works with the ffi module. Here are the files

**reqtest.js**:
```js
console.log("Starting load of module " + process.argv[2])
mod = require(process.argv[2])

console.log(mod)
```

**snapshot.js**:
```js
var NOTHING=1;
```

**package-reqtest.js**:
```js
var nexe = require('nexe');

nexe.compile({
  input: 'reqtest.js',
  output: 'reqtest-bin',
  nodeTempDir: 'node-src',
  nodeConfigureArgs: [
    '--without-dtrace',
    '--without-npm',
    '--without-inspector',
    '--without-etw',
    '--without-perfctr',
  ],
  nodeMakeArgs: ["-j", "8"], // when you want to control the make process.
  nodeVCBuildArgs: ['nosign', 'x64', 'noetw', 'noperfctr'], // when you want to control the make process for windows.
  resourceRoot: 'res',
  flags: true, // use this for applications that need command line flags.
  jsFlags: [ ].join(' '),
  startupSnapshot: 'snapshot.js',
  framework: 'node',
  nodeVersion: '7.4.0',
}, function(err) {
  if (err) { throw err; }
  console.log('Finished bundling. Nexe binary can be found in reqtest-bin');
});
```

So it appears to be something about Lumo or ClojureScript code itself and not the nexe build process that is causing the issue.

**Comment by kanaka** [comment]
Oh, and just FYI, with the following patch to the ref (dep of ffi) module, the ffi module starts loading and working in lumo. It detects if the source object type has a read-only name property and if so it forces the name property of the new object to be writable.

```patch
/node_modules/ref/lib/ref.js 
--- node_modules/ref/lib/ref.js	2017-01-20 09:59:04.323915789 -0600
+++ ../node_modules/ref/lib/ref.js	2017-01-20 10:09:02.431766523 -0600
@@ -306,6 +306,14 @@ exports.refType = function refType (type
   var rtn = Object.create(_type)
   rtn.indirection++
   if (_type.name) {
+    var props = Object.getOwnPropertyDescriptor(_type, "name")
+    if (props && props.writable === false) {
+      props.writable = true
+      Object.defineProperty(rtn, "name", props)
+    }
     rtn.name = _type.name + '*'
   }
   return rtn
```

**Comment by kanaka** [comment]
One final test I did was to make sure that the plain node.js ClojureScript REPL is able to load ffi. 

```
$ cat node_repl.clj
(require 'cljs.repl)
(require 'cljs.repl.node)
(cljs.repl/repl (cljs.repl.node/repl-env) :output-dir "repl-out")

$ rlwrap java -cp cljs.jar:src clojure.main node_repl.clj
ClojureScript Node.js REPL server listening on 52613
To quit, type: :cljs/quit
cljs.user=> (def ffi (js/require "ffi"))
#'cljs.user/ffi
```
That works, so it seems to be something about either the Lumo specific code or the other aspects of the packaging process (the non-nexe parts).

**Comment by kanaka** [comment]
@TooTallNate We are running into a bizarre error with **node-ffi**/**node-ref**. Any chance you could take a look at this issue and give us some insight into what might be happening?

tl;dr version:

* Require of **ffi** using *lumo* (nexe packaged version of ClojureScript) gives this error:
    ```
    TypeError: Cannot assign to read only property 'name' of object '[object Object]'
    ```
* The error comes this [line](https://github.com/TooTallNate/ref/blob/master/lib/ref.js#L309) of the **ref** module (1.3.3) where the `name` property is updated: 
* The error happens for a `[StructType]` with writable flag set to false for the `name` field. However, even when doing this directly from node (7.4.0), the writable property of the `[StructType]` `name` field is also set to false but it works (?!?!).
* [Above is a diff](https://github.com/anmonteiro/lumo/issues/67#issuecomment-274110940) to the **ref** module that makes it work in all cases (by forcing writable for name when it's not)

I've done some bifurcating above and it does appear to be an issue in the lumo code or packaging, but I'm beginning to run out of ideas. Any insight or direction to pursue would be greatly appreciated.

**Comment by kanaka** [comment]
@anmonteiro on a whim, I copied a previously built ClojureScript "snapshot" from target/main.js to snapshot.js and rebuilt my simple test. It's still able to require ffi successfully. So it doesn't appear to be the inclusion of a snapshot during nexe build.

I also just built a version of lumo without the AOT macro bundling and that skips the second packaging step. Lumo gives an error about no such namespace "lumo.repl" but it gives a working prompt and (js/require "ffi") still fails. I think the most likely remaining culprit is the luno code itself.

**Comment by kanaka** [comment]
Aha! Figured it out. The culprit is `v8.setFlagsFromString('--use_strict');` in [src/js/cli.js](https://github.com/anmonteiro/lumo/blob/master/src/js/cli.js#L222)

So we now have the following simple reproducer: `node --use_strict -e "require('ffi')"`

Full example:
```bash
$ npm install ffi
$ node --use_strict -e "require('ffi')"
TypeError: Cannot assign to read only property 'name' of object '[object Object]'
    at Object.refType (.../node_modules/ref/lib/ref.js:318:14)
    at Object.<anonymous> (.../node_modules/ffi/lib/type.js:22:42)
    at Module._compile (module.js:571:32)
    at Object.Module._extensions..js (module.js:580:10)
    at Module.load (module.js:488:32)
    at tryModuleLoad (module.js:447:12)
    at Function.Module._load (module.js:439:3)
    at Module.require (module.js:498:17)
    at require (internal/module.js:20:19)
    at Object.<anonymous> (.../node_modules/ffi/lib/cif.js:6:12)
```

I suppose it makes sense that strict mode would enforce the attribute read/write property. The above patch I suggested to the `ref` module actually seems like less of a hack now that I know what's going on. I'll file a node-ffi bug. Running with node --use_strict seems like a reasonable case where ffi/ref should work.

**Comment by kanaka** [comment]
Pull request sent for ref: https://github.com/TooTallNate/ref/pull/67

**Comment by anmonteiro** [comment]
@kanaka wow, thanks so much for digging into this.

Lumo enables strict mode by default because it just makes everything so much faster.

You can disable it with the following snippet:

```
cljs.user=> (def v8 (js/require "v8"))
#'cljs.user/v8
cljs.user=> (.setFlagsFromString v8 "nouse_strict")
nil
```

Let me know if that makes `node-ffi` work for you.

**Comment by kanaka** [comment]
@anmonteiro yes, that worked, although it looks like it needs to be "--nouse_strict".

**Comment by anmonteiro** [comment]
Looks like this wasn't a bug after all, but a consequence of a Lumo design decision.

**Comment by kanaka** [comment]
Changes to ref module have been committed and a new version of ref pushed to npmjs.org. Given that ffi pulls in version "1" of ref, this should just work now (re-installing ffi/ref module of course).
